BLSQ/iaso
# IA-2935 fix back arrow (TopBar)

Another chunk of the react-router refactor. Get the back arrow in the TopBar to work

Related JIRA tickets : IA-2935

## Self proofreading checklist


## Doc
Not yet

## Changes

- Change `useGoBack`to:
    -  Check the state of the route. If there's a state, it simply goes back one step in the history, else it redirects to the fallback url
    - Take a `nested` prop. For back arrows that redirect to another "back arrow page", it will prevent being unable to navigate back to a page where the whole (hamburger) menu is available
- Change `useRedirections` to pass on the location state when redirecting
- Change `withParams` to use `useParamsObject`.  It will also use the `mappings` url, since that's the only component using it.
- In all `IconButton` with a `url` prop:
    -  prefix url with a `"/"` to force the router to use an absolute path 
    -  pass `location` prop
- Turn methods to generate columns for forms and instances into hooks for clarity
- Add `SUBMISSIONS_UPDATE` to `useCompletenessStatsColumns`
- Cleaned up some components that were still expecting a 'params' prop

## How to test

Before testing:
- update package.json to use [this branch](https://github.com/BLSQ/bluesquare-components/tree/IA-2941_refactor_IconButton) of bluesquare-components and run `npm update bluesquare-components`
- Launch `/api/entityduplicates/analyzes` to populate the entity duplicates page

Testing: 
- Go to all pages with a back arrow in the top bar and go back:
   - They're mostly found by going to a list page (forms, org units etc) and clicking the eye icon on the table
   - Before clicking the icon, add a change in the url, eg: a search (most filters are still broken at this point, so using the input fields won't help)
   - click the eye button
   -  copy the url
   - click the back button: you should go back to the previous page, with the params that you changed 
   - copy the url in a new tab 
   - click the back button: it should redirect to a fallback url. In most cases it's the same url you came from, only the changed params should not be there
 

Expected not to work:

- Completeness stats: it still expects a `router` object, so it crashes
- Mappings: I don't know, I don't have any in my local DB, but aprt from the fact that it's using the `withParams` HOC, it's a pretty straightforward case
- Storages: same as mappings without the HOC
- `LinkToOrgUnit` and `LinkToRegistry`:  still have some redux in them so they won't redirect properly

## Print screen / video

https://github.com/BLSQ/iaso/assets/38907762/a52a8743-ac3b-44a8-a6a2-3a2db1e774ed




## Notes

Depends on https://github.com/BLSQ/bluesquare-components/pull/149 from bluesquare-components. DO NOT MERGE THIS PR EITHER, it is likely to introduce a breaking change in the library
